packages/stokhos/test/UnitTest/Stokhos_SacadoMPVectorUnitTest.cpp

```diff
@@ -474,52 +516,73 @@ struct UnitTestSetup {
+  typedef Stokhos::__storage_type__<int,__scalar_type__,execution_space> storage_type;         \
+  typedef Sacado::MP::Vector<storage_type> vec_type;                                           \
+  typedef UnitTestSetup<vec_type> UTS;                                                         \
+  __macro_for_tests__(__vec_type__,__scalar_type__)
 
-namespace StaticFixedVecTest {
-  typedef Kokkos::DefaultExecutionSpace execution_space;
-  typedef Stokhos::StaticFixedStorage<int,double,8,execution_space> storage_type;
-  typedef Sacado::MP::Vector<storage_type> vec_type;
-  typedef UnitTestSetup<vec_type> UTS;
-  VECTOR_UNIT_TESTS(StaticFixedVector)
+namespace DynamicVecTest
+{
+  TEST_DYNAMIC_STORAGE(DynamicStorage, DynamicVector, double, VECTOR_UNIT_TESTS_REAL_TYPE)
 }
 
-int main( int argc, char* argv[] ) {
-  Teuchos::GlobalMPISession mpiSession(&argc, &argv);
-
-  Kokkos::initialize();
-//  Kokkos::HostSpace::execution_space::initialize();
-//  if (!Kokkos::DefaultExecutionSpace::is_initialized())
-//    Kokkos::DefaultExecutionSpace::initialize();
+namespace DynamicStridedVecTest
+{
+  TEST_DYNAMIC_STORAGE(DynamicStridedStorage, DynamicStridedVector, double, VECTOR_UNIT_TESTS_REAL_TYPE)
+}
 
-  int res = Teuchos::UnitTestRepository::runUnitTestsFromMain(argc, argv);
+/**
+ * Common test structure for static storage.
+ */
+#define TEST_STATIC_STORAGE(__storage_type__,__vec_type__,__scalar_type__,__scalar_type_name__,__storage_size__,__macro_for_tests__)\
+    typedef ::Kokkos::DefaultExecutionSpace execution_space;                                                                            \
+    typedef ::Stokhos::__storage_type__<int,__scalar_type__,__storage_size__,execution_space> storage_type;                             \
+    typedef ::Sacado::MP::Vector<storage_type> vec_type;                                                                                \
+    typedef UnitTestSetup<vec_type> UTS;                                                                                              \
+    __macro_for_tests__(__vec_type__,__scalar_type_name__)
 
-  Kokkos::finalize();
-//  Kokkos::HostSpace::execution_space::finalize();
-//  if (Kokkos::DefaultExecutionSpace::is_initialized())
-//    Kokkos::DefaultExecutionSpace::finalize();
+namespace StaticVecTest
+{
+  TEST_STATIC_STORAGE(StaticStorage, StaticVector, double, double, 8, VECTOR_UNIT_TESTS_REAL_TYPE)
+}
 
-  return res;
+namespace StaticFixedVecTest
+{
+  namespace Double        {TEST_STATIC_STORAGE(StaticFixedStorage, StaticFixedVector,                   double ,                double, 8, VECTOR_UNIT_TESTS_REAL_TYPE   )}
+// Skip std::complex when using CUDA, because std::complex isn't supported in that case.
+#ifndef KOKKOS_ENABLE_CUDA
+  namespace Complex_std   {TEST_STATIC_STORAGE(StaticFixedStorage, StaticFixedVector,   std   ::complex<double>,    std_complex_double, 8, VECTOR_UNIT_TESTS_COMPLEX_TYPE)}
+#endif
```

Comment:
Not sure wether `KOKKOS_ENABLE_CUDA` is the guard to use.  It looks like the code runs on the host only.  So Cuda may not be an issue here.  An issue might be wether a problem arises if `nvcc` is compiling the host code.  If we don't want `nvcc` to compile this line, then we could use `__CUDACC__` as the guard?

Reply:
If I guard with `__CUDACC__` then I don't have the `std::complex` test anymore... shouldn't I use __CUDA_ARCH__?
* https://stackoverflow.com/a/8809924

Reply:
So, since guarding with `__CUDACC__` has the same impact as guarding with `KOKKOS_ENABLE_CUDA`, I propose to use `KOKKOS_ENABLE_CUDA` ;)

Reply:
OK.  The sentence that you added also makes it more clear why we are disabling the line if the compilation is a compilation with cuda.

Reply:
I think you're right, `Kokkos::complex` is always shipped as part of Kokkos. So let's indeed test it whatever the space.